### ISSUE TITLE
Integrate PostHog as a managed app into GitLab

### ISSUE THREAD

**Issue by timgl** [PostHog/posthog#401]
GitLab has asked us to integrate PostHog into GitLab as a managed app. [You can find the issue on GitLab's side here](https://gitlab.com/gitlab-org/gitlab/-/issues/211367#note_309954993)

Initially, this would be as a [CI/CD integration](https://docs.gitlab.com/ee/user/clusters/applications.html#future-apps)

**Comment by timgl** [comment]
## How to create a managed app

[Based on this walkthrough](https://www.youtube.com/watch?v=mKm-jkranEk&feature=youtu.be)

1. [Create a table (migration) for this managed app](https://gitlab.com/gitlab-org/gitlab/-/merge_requests/18015/diffs#3e407d520514a14c5c3b953a2dad7813b66dca97)
2. [Add a model for state management](https://gitlab.com/gitlab-org/gitlab/-/merge_requests/18015/diffs#562199a4378398e686107f9f0f5811a78d596b7f)
 a. This includes helper classes (`Gitlab::Kubernetes::Helm::InstallCommand`) to install the helm chart.
  b. The chart is defined in `def chart` as `stable/elastic-stack`. We probably need to specify a `def repository` where the helm chart is stored
3. **I don't understand if we need to do the frontend or not, or is just having the file in `app/models` enough to be able to do the ci/cd integration** (8:45 in the video)
4. Add a `vendor/posthog/values.yaml` file. This is overridable in the models file
5.  **Specify a CRD?**
6. Upgrading: At the moment only gitlab_runner has a way of being upgraded. They just manually bump version number within GitLab. Another way is to uninstall and re-install the entire managed app. This would be fine but **does it delete the database?**
a. "Chart maintainers aren't building charts in a way that makes them upgradable" (18:15 in the video)

**Comment by kuahyeow** [comment]
@timgl 

The CI/CD integration is actually much simpler. 

Sorry, I haven't managed to write a guide yet. But see https://gitlab.com/gitlab-org/cluster-integration/cluster-applications/-/issues/8 and also https://gitlab.com/gitlab-org/cluster-integration/cluster-applications/-/merge_requests/29 for an example.

You will need:

1. A Helm chart for PostHog
2. A basic understanding of [Helmfile](https://github.com/roboll/helmfile), which is used by our CI/CD integration
2. Create a merge request PostHog using https://gitlab.com/gitlab-org/cluster-integration/cluster-applications/-/merge_requests/29 as an example.

**Comment by mariusandra** [comment]
Hi @kuahyeow and @timgl 

I worked on point 1 from this list and created a helm chart for posthog. It's in PR #407.
I haven't tested running it in any managed kubernetes cluster yet, so I don't know how well it will behave in the real world... or with a managed database. That will be for another day. Locally it seems to run fine though.

Still, any feedback is welcome! :)

**Comment by timgl** [comment]
Thanks @kuahyeow and @mariusandra for the chat!

I've got the following next steps for us:

- [x] Add Ingress resource for Helm chart
- [x] Add a MR to GitLab CI/CD and loop @kuahyeow in
  - To test our MR:
    - Test with helmfile
    - Test with GitLab integrations test
    - [Follow these instructions](https://docs.gitlab.com/ee/user/clusters/applications.html#install-sentry-using-gitlab-cicd)
- [x] Create a MR to GitLab's docs (like Sentry) with sample options
  - https://gitlab.com/gitlab-org/gitlab/-/blob/master/doc/user/clusters/applications.md
  - Look at Jupyter hub for insturctions on how to add secret key
  - Add some instructions for how to do a 'i just want to try it out' setup (with in-cluster database) and 'production setup' (with managed database).

**Comment by mariusandra** [comment]
The latest updates:

- Our helm chart works well for me locally, including ingress.
- I have submitted a MR to Gitlab to include posthog as their managed application, but somehow their tests [don't install it properly](https://gitlab.com/gitlab-org/cluster-integration/cluster-applications/-/merge_requests/74#note_316288472). It could be an issue with the version of helm that they are using (they are on v2, the latest is v3) or something else. I asked and I'm waiting for feedback.
- I created [a draft of the documentation changes](https://gitlab.com/mariusandra/gitlab/-/commit/972de6e88f6318bcb350317f5e81cbf616c9c390). Once the chart is merged, I can make a new MR for the documentation.

**Comment by mariusandra** [comment]
The latest news:
- The [MR](https://gitlab.com/gitlab-org/cluster-integration/cluster-applications/-/merge_requests/74#note_319456602) to add the cluster application is on its way to being merged.
- I created a [MR](https://gitlab.com/gitlab-org/gitlab/-/merge_requests/29152) to add posthog to the gitlab documentation

**Comment by mariusandra** [comment]
Hey, I'd still keep this open. The MR in GitLab is still not merged, so just to not forget about it, let's track the status here.

**Comment by mariusandra** [comment]
I asked when will this get merged: https://gitlab.com/gitlab-org/cluster-integration/cluster-applications/-/merge_requests/74#note_343341529

**Comment by mariusandra** [comment]
Status update.
- [MR merged!](https://gitlab.com/gitlab-org/cluster-integration/cluster-applications/-/merge_requests/74)
- [Documentation MR](https://gitlab.com/gitlab-org/gitlab/-/merge_requests/29152) feedback implemented and its on its way

Comment by Thong on Gitlab:
> FYI, the final piece needed to include PostHog support by default on GitLab is updating the cluster-application image on the GitLab project. This is already happening here: [!32856](https://gitlab.com/gitlab-org/gitlab/-/merge_requests/32856)

Once that chore is done, PostHog is officially in GitLab!

**Comment by mariusandra** [comment]
That PR is merged, so PostHog is officially in the next (released eventually) version of GitLab. In 22 days though, if I understand their milestones correctly: https://gitlab.com/groups/gitlab-org/-/milestones/47

**Comment by mariusandra** [comment]
There were still [docs to rebase](https://gitlab.com/gitlab-org/gitlab/-/merge_requests/29152?diff_id=92389877&start_sha=9b9d62746139c1d71fd1a630c8fa525c766c1a93#note_349212038). Should be merged soon!

**Comment by mariusandra** [comment]
Those are now in. @timgl issue resolved?
GitLab just needs to cut a new release of themselves now and it'll be out in public.
However, they are at PostHog version 1.6.0. To upgrade our release in GitLab we have to:
- Update the `version` in the PostHog helm file, like it was done with [this MR](https://gitlab.com/gitlab-org/cluster-integration/cluster-applications/-/merge_requests/74) [here](https://gitlab.com/gitlab-org/cluster-integration/cluster-applications/-/merge_requests/74/diffs#821ef8281d8cc6b9419dd5b46770a3e78184f0c5_0_16).
- Possibly update the documentation, like it was [done here](https://gitlab.com/gitlab-org/gitlab/-/merge_requests/29152).
- Wait for GitLab to approve and merge those changes
- Wait for GitLab to release a new version of their "Cluster Integration" repo's docker image
- Wait for that new image to be merged into gitlab, like [was done here](https://gitlab.com/gitlab-org/gitlab/-/merge_requests/32856).

Each step requires the review and coordination of many people and I doubt we can get Gitlab to update our app's version every time we make a new release. If we're lucky and really proactive, with this workflow we can get a new posthog version into every other gitlab release. However it will require quite a bit of coordination.
